### ISSUE TITLE
Variant of `-time` flag that prints memory information for every sentence

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#11994]
If I pass `-time` to coqc, I get each line annotated with the time used by that sentence.  I've just learned about `Print Debug GC` (https://github.com/coq/coq/issues/11926#issuecomment-607288150) from @ppedrot, and it'd be great if there was `-mem` or similar to print the change in memory footprint from each sentence (or whichever GC statistic is most useful, or even just all of them).  Bonus points if this can be combined with `-time` to get info about both time and memory (and then I can update various the scripts used with `coq_makefile` to aggregate this information).
It might be nice at the same time to, say, have a setting that makes `Time` and `time` also display memory/GC diffs.

**Comment by ejgallego** [comment]
Note I did implement this in `coq-lsp`, however it doesn't work well at all, suggestions welcome. Problems:

- memory per-sentence is very misleading as sharing is very very large in Coq state; we would need a way to diff among two different states, I'm not sure if there is some facility in OCaml / OCaml libs to do that, maybe Jane Street libs do?
- checking memory use is in general very costly, I see even 20x slowdown in Coq files if I need to call `Obj.reachable_words`

@JasonGross feel free to report wishlist regarding this on `coq-lsp` channels.

**Comment by ppedrot** [comment]
I don't think we need this level of detail. A call to `Gc.quick_stat` should be fast, and that's already an interesting enough piece of data. Furthermore, I personally don't care about sharing, this is irrelevant for efficiency. I just want to see whether Coq is using a lot of memory or not, and if it does I can heat up the memprof tool.

**Comment by JasonGross** [comment]
Yeah, I'm with @ppedrot here.  I wrote a quick-and-dirty script:
```bash
pidf="$(mktemp)"
function cleanup() {
    rm -f "${pidf}"
}
trap cleanup EXIT
rm "${pidf}"; mkfifo "${pidf}"
{ "$@" &
  pidv=$!
  echo $pidv >"$pidf"
  wait $pidv; } | {
    read pid <"$pidf"

    PREV_TIME="$(date +%s.%N)"
    PREV_MEM=0

    while read i
    do
        NEXT_TIME="$(date +%s.%N)"
        NEXT_MEM="$(ps -q "$pid" -o rss 2>/dev/null | tail -1 || echo 0)"
        DIFF_TIME="$(echo "${NEXT_TIME} - ${PREV_TIME}" | bc)"
        DIFF_MEM="$(echo "${NEXT_MEM} - ${PREV_MEM}" | bc)"
        if [ "${DIFF_MEM}" -ge 0 ]; then DIFF_MEM="+${DIFF_MEM}"; fi
        printf "%ss %skb: %s\n" "${DIFF_TIME}" "${DIFF_MEM}" "$i"
        PREV_TIME="${NEXT_TIME}"
        PREV_MEM="${NEXT_MEM}"
    done
}
```
to be invoked as `script.sh coqc -q -time foo.v`, which annotates lines with the amount of RSS they add.  This is already quite helpful

**Comment by ejgallego** [comment]
Ok, I can implement the `Gc.quick_stat` thing in coq-lsp , I assume that for each command you would like to see the diff of the structure?

**Comment by JasonGross** [comment]
Yes

**Comment by ejgallego** [comment]
Done in https://github.com/ejgallego/coq-lsp/pull/61

**Comment by SkySkimmer** [comment]
-profile should have some memory stats too
